### ISSUE TITLE
implement project references

### ISSUE THREAD

**Issue by dagda1** [thefrontside/bigtest#745]
## implementation

currently, every package has its own separate npm/yarn script named `prepack` that builds the current package only.

If any referenced package changes then it is a purely manual process to rebuild the tree of bigtest dependencies.

One solution is to use [typescript project references](https://www.typescriptlang.org/docs/handbook/project-references.html) that is mainly a `tsonfig.json` change with some gotchas due to lack of support from `ts-node` and the bundlers.

To implement project references, we need to update all `tsconfig.json` files to add `"references"` to any dependant packages, e.g. `@bigtest/server` has several references and the `tsconfig.json` would look like this:

```json
{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "outDir": "dist"
  },
  "include": [
    "src/**/*.ts",
    "bin/**/*.ts"
  ],
  "references": [
    { "path": "../agent" },
    { "path": "../atom" },
    { "path": "../bundler" },
    { "path": "../client" },
    { "path": "../driver" },
    { "path": "../effection" },
    { "path": "../effection-express" },
    { "path": "../globals" },
    { "path": "../logging" },
    { "path": "../project" },
    { "path": "../suite" },
    { "path": "../todomvc" },
    { "path": "../webdriver" }
  ]
}
```

each `tsconfig.json` would reference a new base `tsconfig-base.json` that use the `"paths"` option to map the package names to the relevant `src`directory to allow them to be imported like a normal npm package:

```json
{
  "extends": "@frontside/tsconfig",
  "compilerOptions": {
    "baseUrl": ".",
    "composite": true,
    "declaration": true,
    "declarationMap": true,
    "sourceMap": true,
    "paths": {
      "@bigtest/agent": ["packages/agent/src"],
      "@bigtest/atom": ["packages/atom/src"],
      "@bigtest/bigtest": ["packages/bigtest/src"],
      // etc.
```

## gotchas

In order to build a `composite` or a typescript package that has referenced packages `tsc` is called with the `-b` switch.

```bash
tsc -b
```

The downside to `tsc -b` is that no other switches can be supplied.

### ts-node

mocha currently uses `ts-node` to run the tests and ts-node does not support project references

```
mocha -r ts-node/register
```

## bundlers

it does not look like parcel or rollup have adequate support for project references.

**Comment by jbolda** [comment]
I think we could maybe use [tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths) with mocha to support the project references. It hasn't been updated recently, but I imagine this would mostly be an interim change until a start dogfooding?

**Comment by dagda1** [comment]
mocha actually runs fine with ts-node with no changes which is good.

I also noticed a number of `prepack` scripts like this that create commons and esm builds:

```json
"prepack": "tsc --outdir dist/esm --module es2015 && tsc --outdir dist/cjs --module commonjs && tsc --outdir dist/types --emitDeclarationOnly"
```

I personally would use rollup to package these things and not tsc.

I personally would go all-in on rollup and boot out parcel but that is just my opinion.

**Comment by taras** [comment]
@dagda1 what do you think about [esbuild](https://github.com/evanw/esbuild)?

**Comment by dagda1** [comment]
> @dagda1 what do you think about [esbuild](https://github.com/evanw/esbuild)?

it looks interesting, but you never know until you try.  I'm not sure what their web support is like.

If you look at their [FAQ](https://esbuild.github.io/faq/#upcoming-roadmap) they stress a few things that esbuild definitely is not.

**Comment by dagda1** [comment]
closed in #750
